File: weather-dashboard/frontend/src/services/weatherService.ts
import { API } from '../config/api';
import {
    CurrentWeatherData,
    HourlyForecastData,
    DailyForecastData,
    LocationData
} from '../types/weatherTypes';

// Legacy location interface for backward compatibility
export interface Location {
    name: string;
    country: string;
    lat: number;
    lon: number;
}

// Common fields for legacy weather data
interface BaseWeatherData {
    timestamp: number;
    date: string;
    time: string;
}

// Legacy ForecastData structure to maintain compatibility
export interface ForecastData extends BaseWeatherData {
    hour: number;
    temperature: number;
    humidity: number;
    wind_speed: number;
    wind_direction: number;
    description: string;
    icon: string;
    is_day: number | boolean;
    precipitation: number;
}

// Cache management
const memoryCache = new Map();
const CACHE_BUSTER = Date.now().toString();

const getCacheKey = (endpoint: string, lat: number, lon: number) => {
    return `${endpoint}_${lat.toFixed(4)}_${lon.toFixed(4)}_${CACHE_BUSTER}`;
};

const isCacheValid = (cacheTime: number, ttl: number) => {
    return Date.now() - cacheTime < ttl;
};

// Utility functions
const buildApiUrl = (endpoint: string) => `${API.BASE_URL}${endpoint}`;

const handleResponse = async (response: Response) => {
    if (!response.ok) {
        throw new Error(`API error: ${response.status} ${response.statusText}`);
    }
    return await response.json();
};

// Weather code mapping for icon and description
export const fetchWeatherCodes = () => {
    return {
        0: { description: "Clear sky", icon: "clear" },
        1: { description: "Mainly clear", icon: "mainly-clear" },
        2: { description: "Partly cloudy", icon: "partly-cloudy" },
        3: { description: "Overcast", icon: "cloudy" },
        45: { description: "Fog", icon: "fog" },
        48: { description: "Depositing rime fog", icon: "fog" },
        51: { description: "Light drizzle", icon: "drizzle" },
        53: { description: "Moderate drizzle", icon: "drizzle" },
        55: { description: "Dense drizzle", icon: "drizzle" },
        56: { description: "Light freezing drizzle", icon: "freezing-drizzle" },
        57: { description: "Dense freezing drizzle", icon: "freezing-drizzle" },
        61: { description: "Slight rain", icon: "rain" },
        63: { description: "Moderate rain", icon: "rain" },
        65: { description: "Heavy rain", icon: "heavy-rain" },
        66: { description: "Light freezing rain", icon: "freezing-rain" },
        67: { description: "Heavy freezing rain", icon: "freezing-rain" },
        71: { description: "Slight snow fall", icon: "snow" },
        73: { description: "Moderate snow fall", icon: "snow" },
        75: { description: "Heavy snow fall", icon: "heavy-snow" },
        77: { description: "Snow grains", icon: "snow-grains" },
        80: { description: "Slight rain showers", icon: "rain-showers" },
        81: { description: "Moderate rain showers", icon: "rain-showers" },
        82: { description: "Violent rain showers", icon: "heavy-rain-showers" },
        85: { description: "Slight snow showers", icon: "snow-showers" },
        86: { description: "Heavy snow showers", icon: "heavy-snow-showers" },
        95: { description: "Thunderstorm", icon: "thunderstorm" },
        96: { description: "Thunderstorm with slight hail", icon: "thunderstorm-hail" },
        99: { description: "Thunderstorm with heavy hail", icon: "thunderstorm-hail" }
    };
};

// Generic fetch handler with error management and JSON parsing
const fetchWithErrorHandling = async (url: string): Promise<any> => {
    try {
        console.log(`Fetching: ${url}`);
        const response = await fetch(url);

        if (!response.ok) {
            throw new Error(`HTTP error! Status: ${response.status}`);
        }

        return await response.json();
    } catch (error) {
        console.error(`Error fetching data from ${url}:`, error);
        throw error;
    }
};

// Fetch current weather data for a given location
export const fetchCurrentWeather = async (latitude: number, longitude: number): Promise<CurrentWeatherData> => {
    const url = `${API.baseUrl}/current?latitude=${latitude}&longitude=${longitude}`;
    const data = await fetchWithErrorHandling(url);

    // Validate and transform the data to match the interface
    return {
        temperature_2m: data.temperature || 0,
        apparent_temperature: data.apparent_temperature || 0,
        feels_like_temperature: data.feels_like_temperature || data.apparent_temperature || 0,
        relative_humidity_2m: data.relative_humidity || data.humidity || 0,
        precipitation: data.precipitation || 0,
        precipitation_probability: data.precipitation_probability || 0,
        wind_speed_10m: data.wind_speed || 0,
        wind_direction_10m: data.wind_direction || 0,
        surface_pressure: data.surface_pressure || 1013,
        weather_code: data.weather_code || 0,
        is_day: data.is_day !== undefined ? data.is_day : 1,
        uv_index: data.uv_index || 0,
        cloud_cover: data.cloud_cover || 0
    };
};

// Fetch hourly forecast data for a given location
export const fetchHourlyForecast = async (
    latitude: number,
    longitude: number,
    hours: number = 24
): Promise<HourlyForecastData> => {
    const url = `${API.baseUrl}/hourly?latitude=${latitude}&longitude=${longitude}&hours=${hours}`;
    const data = await fetchWithErrorHandling(url);

    // Ensure we return a properly formatted data object
    return {
        timestamps: data.timestamps || data.time || [],
        temperature_2m: data.temperature_2m || data.temperature || [],
        precipitation: data.precipitation || [],
        precipitation_probability: data.precipitation_probability || [],
        weather_code: data.weather_code || [],
        wind_speed_10m: data.wind_speed_10m || data.wind_speed || [],
        wind_direction_10m: data.wind_direction_10m || data.wind_direction || [],
        is_day: data.is_day || []
    };
};

// Fetch daily forecast data for a given location
export const fetchDailyForecast = async (
    latitude: number,
    longitude: number,
    days: number = 7
): Promise<DailyForecastData> => {
    const url = `${API.baseUrl}/daily?latitude=${latitude}&longitude=${longitude}&days=${days}`;
    const data = await fetchWithErrorHandling(url);

    // Ensure we return a properly formatted data object
    return {
        time: data.time || data.dates || [],
        temperature_2m_max: data.temperature_2m_max || data.temperature_max || [],
        temperature_2m_min: data.temperature_2m_min || data.temperature_min || [],
        precipitation_sum: data.precipitation_sum || data.precipitation || [],
        precipitation_probability_max: data.precipitation_probability_max ||
            data.precipitation_probability || [],
        weather_code: data.weather_code || [],
        wind_speed_10m_max: data.wind_speed_10m_max || data.wind_speed_max || [],
        wind_direction_10m_dominant: data.wind_direction_10m_dominant || [],
        sunrise: data.sunrise || [],
        sunset: data.sunset || []
    };
};

// Search for a location by query string
export const fetchLocationBySearch = async (query: string): Promise<LocationData> => {
    const url = `${API.geocodingUrl}?query=${encodeURIComponent(query)}`;
    const data = await fetchWithErrorHandling(url);

    if (!data || !data.latitude || !data.longitude) {
        throw new Error('Location not found');
    }

    return {
        name: data.name || 'Unknown Location',
        country: data.country || '',
        state: data.state || '',
        latitude: data.latitude,
        longitude: data.longitude
    };
};

export const clearWeatherCache = () => {
    // Clear only weather-related cache entries
    const keysToDelete = [];
    for (const key of Array.from(memoryCache.keys())) {
        if (key.includes(API.ENDPOINTS.CURRENT) ||
            key.includes(API.ENDPOINTS.HOURLY) ||
            key.includes(API.ENDPOINTS.DAILY)) {
            keysToDelete.push(key);
        }
    }

    keysToDelete.forEach(key => memoryCache.delete(key));
    console.log('Weather cache cleared');
};

File: backend/weather_service.py
from datetime import date, datetime, timedelta
import requests
from typing import Dict, Any, List, Optional

# Base URL for the Open-Meteo API
OPEN_METEO_URL = "https://api.open-meteo.com/v1"

def fetch_current_weather(latitude: float, longitude: float) -> Dict[str, Any]:
    """Fetch current weather data for a specific location"""

    endpoint = f"{OPEN_METEO_URL}/forecast"

    params = {
        "latitude": latitude,
        "longitude": longitude,
        "current": [
            "temperature_2m",
            "apparent_temperature",
            "relative_humidity_2m",
            "precipitation",
            "weather_code",
            "wind_speed_10m",
            "wind_direction_10m",
            "surface_pressure",
            "is_day"
        ],
        "timezone": "auto"
    }

    response = requests.get(endpoint, params=params)
    response.raise_for_status()
    data = response.json()

    # Transform the response to match frontend's expected format
    current = data.get("current", {})

    return {
        "temperature_2m": current.get("temperature_2m"),
        "apparent_temperature": current.get("apparent_temperature"),
        "relative_humidity_2m": current.get("relative_humidity_2m"),
        "precipitation": current.get("precipitation"),
        "weather_code": current.get("weather_code"),
        "wind_speed_10m": current.get("wind_speed_10m"),
        "wind_direction_10m": current.get("wind_direction_10m"),
        "surface_pressure": current.get("surface_pressure"),
        "is_day": current.get("is_day"),
        "timestamp": current.get("time")
    }

def fetch_hourly_forecast(latitude: float, longitude: float, hours: int = 24) -> Dict[str, Any]:
    """Fetch hourly forecast data for a specific location"""

    endpoint = f"{OPEN_METEO_URL}/forecast"

    params = {
        "latitude": latitude,
        "longitude": longitude,
        "hourly": [
            "temperature_2m",
            "apparent_temperature",
            "precipitation_probability",
            "precipitation",
            "rain",
            "showers",
            "snowfall",
            "cloud_cover",
            "weather_code",
            "wind_speed_10m",
            "wind_direction_10m",
            "relative_humidity_2m",
            "wind_gusts_10m",
            "is_day"
        ],
        "forecast_hours": hours,
        "timezone": "auto"
    }

    response = requests.get(endpoint, params=params)
    response.raise_for_status()
    data = response.json()

    # Transform the response to match frontend's expected format
    hourly = data.get("hourly", {})
    times = hourly.get("time", [])[:hours]

    return {
        "timestamps": times,
        "temperature_2m": hourly.get("temperature_2m", [])[:hours],
        "apparent_temperature": hourly.get("apparent_temperature", [])[:hours],
        "precipitation_probability": hourly.get("precipitation_probability", [])[:hours],
        "precipitation": hourly.get("precipitation", [])[:hours],
        "rain": hourly.get("rain", [])[:hours],
        "showers": hourly.get("showers", [])[:hours],
        "snowfall": hourly.get("snowfall", [])[:hours],
        "cloud_cover": hourly.get("cloud_cover", [])[:hours],
        "weather_code": hourly.get("weather_code", [])[:hours],
        "wind_speed_10m": hourly.get("wind_speed_10m", [])[:hours],
        "wind_direction_10m": hourly.get("wind_direction_10m", [])[:hours],
        "relative_humidity_2m": hourly.get("relative_humidity_2m", [])[:hours],
        "wind_gusts_10m": hourly.get("wind_gusts_10m", [])[:hours],
        "is_day": hourly.get("is_day", [])[:hours]
    }

def fetch_historical_weather(
    latitude: float,
    longitude: float,
    start_date: date,
    end_date: date
) -> Dict[str, Any]:
    """Fetch historical weather data for a specific location and date range"""

    # Format dates as strings for the API
    start_date_str = start_date.strftime("%Y-%m-%d")
    end_date_str = end_date.strftime("%Y-%m-%d")

    endpoint = f"{OPEN_METEO_URL}/forecast"  # Changed to forecast endpoint for daily forecast

    params = {
        "latitude": latitude,
        "longitude": longitude,
        "daily": [
            "temperature_2m_max",
            "temperature_2m_min",
            "precipitation_sum",
            "precipitation_probability_max",
            "wind_speed_10m_max",
            "wind_direction_10m_dominant",
            "weather_code"
        ],
        "timezone": "auto",
        "start_date": start_date_str,
        "end_date": end_date_str
    }

    response = requests.get(endpoint, params=params)
    response.raise_for_status()
    data = response.json()

    # Transform the response to match frontend's expected format
    daily = data.get("daily", {})

    return {
        "time": daily.get("time", []),
        "temperature_2m_max": daily.get("temperature_2m_max", []),
        "temperature_2m_min": daily.get("temperature_2m_min", []),
        "precipitation_sum": daily.get("precipitation_sum", []),
        "precipitation_probability_max": daily.get("precipitation_probability_max", []),
        "wind_speed_10m_max": daily.get("wind_speed_10m_max", []),
        "wind_direction_10m_dominant": daily.get("wind_direction_10m_dominant", []),
        "weather_code": daily.get("weather_code", [])
    }